### PR TITLE
lazy load adapaters, thread safe

### PR DESCRIPTION
**Pull request by jeffdaily**
Follow-up to PR #385 and #532.

### CODE REVIEW

**Comment by jeffdaily** [comment]
This PR starts from recently merged #385 but adapts the hotfix from rocm-rel-6.0 branch to make it thread safe.  I found this PR diff easier to review in github if I turned off whitespace changes.

Also, made the changes without attempting a build, so this does need a CI signal before merging.

**Review by Serge45 — APPROVED** [review]

**Comment by jeffdaily** [comment]
This really needs to be tested on MI100 and some navi system before landing.  So let's not merge until this change can be staged appropriately.

**Comment by jichangjichang** [comment]
close old PR

**Pull request closed without merging**
